AttributeError: '_TroveClassifier' object has no attribute '__name__'

```
.tox/default/lib/python3.8/site-packages/validate_pyproject/api.py:67: in <dictcomp>
    if callable(fn) and not fn.__name__.startswith("_")
E   AttributeError: '_TroveClassifier' object has no attribute '__name__'
```

--- a/src/validate_pyproject/api.py
+++ b/src/validate_pyproject/api.py
@@ -7,7 +7,7 @@ import sys
 from enum import Enum
 from functools import partial, reduce
 from itertools import chain
-from types import MappingProxyType
+from types import MappingProxyType, ModuleType
 from typing import (
     TYPE_CHECKING,
     Callable,
@@ -28,6 +28,9 @@ from fastjsonschema.ref_resolver import RefResolver
 from . import errors, formats
 from .extra_validations import EXTRA_VALIDATIONS
 from .types import FormatValidationFn, Schema, ValidationFn
+
+_logger = logging.getLogger(__name__)
+_chain_iter = chain.from_iterable
 
 if TYPE_CHECKING:  # pragma: no cover
     from .plugins import PluginWrapper  # noqa
@@ -60,16 +63,16 @@ ALL_PLUGINS = AllPlugins.ALL_PLUGINS
 TOP_LEVEL_SCHEMA = "pyproject_toml"
 PROJECT_TABLE_SCHEMA = "project_metadata"
 
-FORMAT_FUNCTIONS: Mapping[str, FormatValidationFn] = MappingProxyType(
-    {
+
+def _get_public_functions(module: ModuleType) -> Mapping[str, FormatValidationFn]:
+    return {
         fn.__name__.replace("_", "-"): fn
-        for fn in formats.__dict__.values()
+        for fn in module.__dict__.values()
         if callable(fn) and not fn.__name__.startswith("_")
     }
-)
 
-_logger = logging.getLogger(__name__)
-_chain_iter = chain.from_iterable
+
+FORMAT_FUNCTIONS = MappingProxyType(_get_public_functions(formats))
 
 
 def load(name: str, package: str = __package__, ext: str = ".schema.json") -> Schema:

--- a/src/validate_pyproject/formats.py
+++ b/src/validate_pyproject/formats.py
@@ -133,6 +133,7 @@ class _TroveClassifier:
         self.downloaded: typing.Union[None, False, typing.Set[str]] = None
         # None => not cached yet
         # False => cache not available
+        self.__name__ = "trove_classifier"  # Emulate a public function
 
     def __call__(self, value: str) -> bool:
         if self.downloaded is False:

--- a/tests/test_formats.py
+++ b/tests/test_formats.py
@@ -5,6 +5,7 @@ from unittest.mock import Mock
 
 import pytest
 
+from validate_pyproject import api
 from validate_pyproject import formats
 
 _chain_iter = chain.from_iterable
@@ -284,6 +285,33 @@ class TestClassifiers:
         "Operating System :: POSIX",
         "Programming Language :: Python :: 3 :: Only",
     )
+
+    def test_does_not_break_public_function_detection(self):
+        # See https://github.com/abravalheri/validate-pyproject/issues/12
+
+        # When `trove_classifiers` is defined from the dependency package
+        # it will be a function.
+        # When it is defined based on the download, it will be a custom object.
+
+        # In both cases the `_TroveClassifier` class should not be made public,
+        # but the instance should
+
+        trove_classifier = formats._TroveClassifier()
+        _formats = Mock(
+            _TroveClassifier=formats._TroveClassifier,
+            trove_classifiers=trove_classifier,
+        )
+        fns = api._get_public_functions(_formats)
+        assert fns == {"trove-classifier": trove_classifier}
+
+        # Make sure the object and the function have the same name
+        assert "trove-classifier" in api.FORMAT_FUNCTIONS
+        normalized_name = trove_classifier.__name__.replace("_", "-")
+        assert normalized_name == "trove-classifier"
+        assert normalized_name in api.FORMAT_FUNCTIONS
+
+        func_name = trove_classifier.__name__
+        assert getattr(formats, func_name) in api.FORMAT_FUNCTIONS.values()
 
     def test_download(self):
         try:

--- a/tests/test_formats.py
+++ b/tests/test_formats.py
@@ -5,8 +5,7 @@ from unittest.mock import Mock
 
 import pytest
 
-from validate_pyproject import api
-from validate_pyproject import formats
+from validate_pyproject import api, formats
 
 _chain_iter = chain.from_iterable
 

